src/protocols/bdx/tests/TestBdxTransferSession.cpp

```diff
@@ -0,0 +1,335 @@
+    NL_TEST_ASSERT(inSuite, outEvent.transferAcceptData.Length == proposedLength);
+    NL_TEST_ASSERT(inSuite, outEvent.transferAcceptData.MaxBlockSize == transferBlockSize);
+    System::PacketBufferHandle rcvdAcceptMsgData = outEvent.MsgData.Retain();
+    NL_TEST_ASSERT(inSuite, !rcvdAcceptMsgData.IsNull());
+    NL_TEST_ASSERT(inSuite,
+                   !memcmp(testMetadata, outEvent.transferAcceptData.Metadata, outEvent.transferAcceptData.MetadataLength));
+
+    // Test BlockQuery -> Block -> BlockAck
+    SendAndVerifyQuery(inSuite, inContext, respondingSender, initiatingReceiver, outEvent);
+    SendAndVerifyArbitraryBlock(inSuite, inContext, respondingSender, initiatingReceiver, outEvent, false);
+    numBlocksSent++;
+    SendAndVerifyBlockAck(inSuite, inContext, respondingSender, initiatingReceiver, outEvent);
+
+    // Test multiple Blocks sent and received (last Block is BlockEOF)
+    while (numBlocksSent < numBlockSends)
+    {
+        bool isEof = (numBlocksSent == numBlockSends - 1);
+
+        SendAndVerifyQuery(inSuite, inContext, respondingSender, initiatingReceiver, outEvent);
+        SendAndVerifyArbitraryBlock(inSuite, inContext, respondingSender, initiatingReceiver, outEvent, isEof);
+
+        numBlocksSent++;
+    }
+
+    // Verify last block was BlockEOF, then verify that PrepareBlockAckEOF() emits BlockAckEOF message
+    NL_TEST_ASSERT(inSuite, outEvent.blockdata.IsEof == true);
+    err = initiatingReceiver.PrepareBlockAckEOF();
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    initiatingReceiver.PollOutput(outEvent);
+    NL_TEST_ASSERT(inSuite, outEvent.EventType == TransferSession::kOutput_MsgToSend);
+    System::PacketBufferHandle eofAckToSend = outEvent.MsgData.Retain();
+    NL_TEST_ASSERT(inSuite, !eofAckToSend.IsNull());
+
+    // Pass BlockAckEOF message to respondingSender and verify PollOutput emits correct event
+    err = respondingSender.HandleMessageReceived(eofAckToSend.Retain());
+    NL_TEST_ASSERT(inSuite, err == CHIP_NO_ERROR);
+    respondingSender.PollOutput(outEvent);
+    NL_TEST_ASSERT(inSuite, outEvent.EventType == TransferSession::kOutput_AckEOFReceived);
+}
+
+// Partial transfer test using Sender Drive to specifically test Block -> BlockAck -> Block sequence
+void TestInitiatingSenderSenderDrive(nlTestSuite * inSuite, void * inContext)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    TransferSession::OutputEvent outEvent;
+    TransferSession initiatingSender;
+    TransferSession respondingReceiver;
+
+    TransferControlFlags driveMode = kControl_SenderDrive;
+
+    // Chosen arbitrarily for this test
+    uint16_t transferBlockSize = 10;
+
+    // Initialize struct with TransferInit parameters
+    TransferSession::TransferInitData initOptions;
+    initOptions.TransferCtlFlagsRaw = driveMode;
+    initOptions.MaxBlockSize        = transferBlockSize;
+    char testFileDes[9]             = { "test.txt" };
+    initOptions.FileDesLength       = 9;
```

Comment:
Same comment about file des length NOT including the nul terminator.

Reply:
ack